e3db/tsv1_auth.py

```diff
@@ -0,0 +1,47 @@
+from e3db.tests.test_tsv1_auth import PUBLIC_KEY_BYTES
+from nacl import public
+import requests
+from requests import auth
+from requests.auth import AuthBase
+from .sodium_crypto import SodiumCrypto
+from .base_crypto import BaseCrypto
+import datetime 
+
+class E3DBTSV1Auth(AuthBase):
+    HASHING_ALGORITHM = "BLAKE2B"
+    SIGNATURE_TYPE = "ED25519"
+    AUTHENTICATION_METHOD = "TSV1-" + SIGNATURE_TYPE + "-" + HASHING_ALGORITHM
+    AUTHORIZATION_HEADER = "Authorization"
+    SECRET_KEY_BYTES = 32
+
+    def __init__(self, private_signing_key: str, client_id: str):
+        self.private_signing_key = private_signing_key
+        self.client_id = client_id
+
+    def __call__(self, r: requests.Request):
+        """
+        TSV1 Authentication
+        """
+        private_b64_decoded = BaseCrypto.base64decode(self.private_signing_key)[:self.SECRET_KEY_BYTES]
+        signing_key = SodiumCrypto.generate_signing_key(private_b64_decoded)
+        public_signing_key = signing_key.verify_key
+        public_b64 = BaseCrypto.base64encode(public_signing_key).decode("utf-8")
+
+        timestamp = datetime.datetime.now()
+        nonce = SodiumCrypto.random_nonce()
+
+        header_string = f"{self.AUTHENTICATION_METHOD}; {public_b64}; {timestamp}; {nonce}; uid:{self.client_id}"
+
+        call_path = r.url
+        query_string = r.params # Double check sortedness of params
```

Comment:
what should the sort order be and can we add a helpful comment here to make the next developer's (including you in 3 months when ;-) comprehension higher

Reply:
This should use the sort order for TSV1 which is based on https://docs.aws.amazon.com/general/latest/gr/sigv4-create-canonical-request.html Step 3.a

Reply:
https://github.com/tozny/e3db-java/blob/1a66de30308f1eacc436c03095e19281b96c5f58/e3db/src/main/java/com/tozny/e3db/Client.java#L987 This is the java implementation of query param string building